audioipc/src/rpccore.rs

```diff
@@ -100,25 +111,64 @@ impl<Request, Response> Proxy<Request, Response> {
 
 impl<Request, Response> Clone for Proxy<Request, Response> {
     fn clone(&self) -> Self {
-        Proxy {
+        let (tx, rx) = crossbeam_channel::bounded(1);
+        let key = self.proxy_mgr.upgrade().unwrap().register_proxy(tx);
+        Self {
             handle: self.handle.clone(),
-            tx: self.tx.clone(),
+            key,
+            response_rx: rx,
+            handler_tx: self.handler_tx.clone(),
+            proxy_mgr: self.proxy_mgr.clone(),
         }
     }
 }
 
 impl<Request, Response> Drop for Proxy<Request, Response> {
     fn drop(&mut self) {
         trace!("Proxy drop, waking EventLoop");
+        if let Some(mgr) = self.proxy_mgr.upgrade() {
+            mgr.unregister_proxy(self.key)
+        }
         // Must drop Sender before waking the connection, otherwise
         // the wake may be processed before Sender is closed.
-        unsafe { ManuallyDrop::drop(&mut self.tx) }
+        unsafe {
+            ManuallyDrop::drop(&mut self.handler_tx);
+        }
         if self.handle.is_some() {
             self.wake_connection()
         }
     }
 }
 
+#[derive(Debug)]
+struct ProxyManager<Response> {
+    proxies: Mutex<Slab<Sender<Response>>>,
+}
```

Comment:
Would you mind adding some comments explaining the ownership of the `ProxyManager`? It should be able to help explain when `upgrade()` works.

It's owned by `ClientHandler`, which will be tied to `Connection` and then `EventLoop` eventually and so it will live longer than `Proxy`. As a result, `proxy_mgr.upgrade()` are always expected to work?

Reply:
Add comments explaining this.